### ISSUE TITLE
Duplicate machine name in vRA unless parallelism=1

### ISSUE THREAD

**Issue by jasonlivie** [vmware-archive/legacy-terraform-provider-vra7#124]
**vRA 7.x version**
vRA 7.4

**Terraform version**
Terraform v0.11.11

**terraform-provider-vra7 plugin version**

9c6face874005ba2cea76ed7ef0c5bcc87cccf33

**Describe the bug**

When trying to create multiple VMs, vRA fails with *Error requesting machine. Machine name <hostname> already exists* unless -parallelism=1. 

**To Reproduce**
1. Terraform config file

```
resource "vra7_resource" "worker_vm1" {
  wait_timeout = 120

  catalog_name = "Worker"

  deployment_configuration = {
    description = "Deployed Worker Class with Terraform vm1"
  }

  resource_configuration = {
    vSphere__vCenter__Machine_1.memory =  "${var.vm_memory}"
    vSphere__vCenter__Machine_1.description = "Deployed Worker with Terraform"


    vSphere__vCenter__Machine_1.Hostname = "vmtest1"
    vSphere__vCenter__Machine_1.VMware.VirtualCenter.OperatingSystem = "${var.vm_OperatingSystem}"
    vSphere__vCenter__Machine_1.VirtualMachine.Network0.Name = "${var.vm_network0_name}"
    vSphere__vCenter__Machine_1.COMPANY.Domain.Name = "${var.vm_domain_name}"

    vSphere__vCenter__Machine_1.ReservationID_Test = "${var.vm_res_pol_id}"

    vSphere__vCenter__Machine_1.COMPANY.Integrator = "${var.req_integrator}"
    vSphere__vCenter__Machine_1.COMPANY.Project.Code = "${var.req_project_code}"
    vSphere__vCenter__Machine_1.COMPANY.Requested.By = "${var.req_project_manager}"
    vSphere__vCenter__Machine_1.COMPANY.PTC.Request = "${var.req_ptc}"
    vSphere__vCenter__Machine_1.COMPANY.DR.Type= "${var.dr_type}"

    vSphere__vCenter__Machine_1.COMPANY.Satelitte.Location = "${var.sat_location}"
    vSphere__vCenter__Machine_1.COMPANY.SatelitteHostgroups = "${var.hg}"
    vSphere__vCenter__Machine_1.COMPANY.SatelitteOrganizationPTC = "${var.sat_org}"
  }
}

resource "vra7_resource" "worker_vm2" {
  wait_timeout = 120

  catalog_name = "Worker"

  deployment_configuration = {
    description = "Deployed Worker Class with Terraform vm2"
  }

  resource_configuration = {
    vSphere__vCenter__Machine_1.memory =  "${var.vm_memory}"
    vSphere__vCenter__Machine_1.description = "Deployed Worker with Terraform"


    vSphere__vCenter__Machine_1.Hostname = "vmtest2"
    vSphere__vCenter__Machine_1.VMware.VirtualCenter.OperatingSystem = "${var.vm_OperatingSystem}"
    vSphere__vCenter__Machine_1.VirtualMachine.Network0.Name = "${var.vm_network0_name}"
    vSphere__vCenter__Machine_1.COMPANY.Domain.Name = "${var.vm_domain_name}"

    vSphere__vCenter__Machine_1.ReservationID_Test = "${var.vm_res_pol_id}"

    vSphere__vCenter__Machine_1.COMPANY.Integrator = "${var.req_integrator}"
    vSphere__vCenter__Machine_1.COMPANY.Project.Code = "${var.req_project_code}"
    vSphere__vCenter__Machine_1.COMPANY.Requested.By = "${var.req_project_manager}"
    vSphere__vCenter__Machine_1.COMPANY.PTC.Request = "${var.req_ptc}"
    vSphere__vCenter__Machine_1.COMPANY.DR.Type= "${var.dr_type}"

    vSphere__vCenter__Machine_1.COMPANY.Satelitte.Location = "${var.sat_location}"
    vSphere__vCenter__Machine_1.COMPANY.SatelitteHostgroups = "${var.hg}"
    vSphere__vCenter__Machine_1.COMPANY.SatelitteOrganizationPTC = "${var.sat_org}"
  }
}
```

2. Terraform command
```
terraform apply <= fails
terraform apply -parallelism=1 <= works
```
3. Error

`Error requesting machine. Machine name <hostname> already exists` in vRA

According to the debug logs of terraform,  both resources (vm hostnames) exists as separate entities and are distinct. This is found in the `Resource Configuration: map[..]` section of the logs . 

Now when the `Updated template - map[..]`  is generated for both resources, we get duplicate hostnames in the structure. The JSON payload is therefore sent with duplicate hostsnames for both resources

If we are to use parallelism=1, the the `Update template -map` and JSON payload are generated with 2 distinct hostnames

**Logs**

Please let me know if debug logs are required.  I can always add snippets of logs. This is ran at a client location, and major cleanup would be required.

Thanks for your help.

**Comment by jonnyborbs** [comment]
Hi @jasonlivie  - thanks for the detail here. Can you provide some more info on the reproduction steps, environment and answer a few quick questions? It would be helpful if you could:

- Export the blueprint you're using within vRA and attach it here
- Give some details about the template OS/configuration you're trying to deploy

Also, can I infer from your variable names that you are integrating in some way with Red Hat Satellite here? More info about what is going on there would be helpful.

One final thought - I notice that you used `vSphere__vCenter__Machine_1.` in both resources, despite them being created as worker_vm1 and worker_vm2. What happens if you change that object identifier to `vSphere__vCenter__Machine_2.` in the worker_vm2 definition? Just curious.

**Comment by jasonlivie** [comment]
Hi @vaficionado 

When deploying VMs, we are requesting them with different hostnames (and possibly other values), yet when we look in the vRA UI, both requests come in with the same values. Instead of seeing 2 distinct requests, we see 1 request being sent twice. This is why our requests are failing with a duplicate machine name



* Please find the blueprint attached
* We are deploying a single machine blueprint where the VM is identified by “vSphere__vCenter__Machine_1”, we cannot change to “_2”, that won’t work
* Satellite comes into play much later in vRO, the issue is with duplicate VM names in this case
* We are deploying Red Hat 7 OS VM (rhel7_64Guest) that is deployed with Satellite via vRO but we don’t get to this point.

[Blueprint-Worker.yaml](https://github.com/vmware/terraform-provider-vra7/files/2886555/Blueprint-Worker.yaml.log)

thanks

**Comment by jonnyborbs** [comment]
Very interesting. @Prativa20 have you seen this before?

**Comment by Prativa20** [comment]
@vaficionado Yes, I could reproduce this issue. When I requested two resources with different configuration, 2 exact same deployments were created. This is a bug. Looking into it.

Thanks,
Prativa

**Comment by mponton** [comment]
@Prativa20 I work with @jasonlivie and was looking at the code, I think the issue is that all of these:
```go
var (
	log                     = logging.MustGetLogger(utils.LoggerID)
	catalogItemName         string
	catalogItemID           string
	businessGroupID         string
	businessGroupName       string
	waitTimeout             int
	requestStatus           string
	failedMessage           string
	deploymentConfiguration map[string]interface{}
	resourceConfiguration   map[string]interface{}
	catalogConfiguration    map[string]interface{}
)
```
Are globals inside the `vrealize` module.

When more than one request comes in, values are overwritten:
```
2019-02-26T13:47:55.904-05:00 resource.go:805 readProviderConfiguration ▶ INFO 001 Reading the provider configuration data.....
2019-02-26T13:47:55.906-05:00 resource.go:807 readProviderConfiguration ▶ INFO 002 Catalog item name: Red Hat Linux 
2019-02-26T13:47:55.906-05:00 resource.go:809 readProviderConfiguration ▶ INFO 003 Catalog item ID: 
2019-02-26T13:47:55.906-05:00 resource.go:811 readProviderConfiguration ▶ INFO 004 Business Group name: 
2019-02-26T13:47:55.906-05:00 resource.go:813 readProviderConfiguration ▶ INFO 005 Business Group Id: 
2019-02-26T13:47:55.906-05:00 resource.go:815 readProviderConfiguration ▶ INFO 006 Wait time out: 7200 
2019-02-26T13:47:55.906-05:00 resource.go:817 readProviderConfiguration ▶ INFO 007 Failed message:  
2019-02-26T13:47:55.906-05:00 resource.go:819 readProviderConfiguration ▶ INFO 008 Resource Configuration: REMOVED
2019-02-26T13:47:55.906-05:00 resource.go:821 readProviderConfiguration ▶ INFO 009 Deployment Configuration: map[description:Deployed with Terraform] 
2019-02-26T13:47:55.906-05:00 resource.go:823 readProviderConfiguration ▶ INFO 00a Catalog configuration: map[] 
2019-02-26T13:47:55.906-05:00 catalog.go:93 readCatalogItemIDByName ▶ INFO 00b readCatalogItemIdByName->catalog_name Red Hat Linux
2019-02-26T13:47:55.91-05:00 resource.go:805 readProviderConfiguration ▶ INFO 00c Reading the provider configuration data.....
2019-02-26T13:47:55.911-05:00 resource.go:807 readProviderConfiguration ▶ INFO 00d Catalog item name: Red Hat Linux 
2019-02-26T13:47:55.911-05:00 resource.go:809 readProviderConfiguration ▶ INFO 00e Catalog item ID: 
2019-02-26T13:47:55.911-05:00 resource.go:811 readProviderConfiguration ▶ INFO 00f Business Group name: 
2019-02-26T13:47:55.911-05:00 resource.go:813 readProviderConfiguration ▶ INFO 010 Business Group Id: 
2019-02-26T13:47:55.911-05:00 resource.go:815 readProviderConfiguration ▶ INFO 011 Wait time out: 7200 
2019-02-26T13:47:55.911-05:00 resource.go:817 readProviderConfiguration ▶ INFO 012 Failed message:  
2019-02-26T13:47:55.911-05:00 resource.go:819 readProviderConfiguration ▶ INFO 013 Resource Configuration: REMOVED
2019-02-26T13:47:55.911-05:00 resource.go:821 readProviderConfiguration ▶ INFO 014 Deployment Configuration: map[description:Deployed with Terraform] 
2019-02-26T13:47:55.911-05:00 resource.go:823 readProviderConfiguration ▶ INFO 015 Catalog configuration: map[] 
2019-02-26T13:47:55.911-05:00 catalog.go:93 readCatalogItemIDByName ▶ INFO 016 readCatalogItemIdByName->catalog_name Red Hat Linux
```
In the above, when the line ...
```
2019-02-26T13:47:55.91-05:00 resource.go:805 readProviderConfiguration ▶ INFO 00c Reading the provider configuration data.....
```
...starts, it overwrites previous values before any request is even submitted.

Unless I am mistaken, I see two options to fix:
1. A somewhat important refactor to have `readProviderConfiguration` return a structure with all the `vra7_resource` values that would be used in calls to other functions (instead of globals).
2. Add a lock at the start of `createResource` and that is released before `return waitForRequestCompletion(d, meta)`.

The latter is surely a lot less work. It would block parallel requests and process a single request at once but only until the request has been submitted to the vRA API. A good compromise I think.

That said, I did not have the time fully analyze if we'd encounter other issues doing that.

**Comment by markpeek** [comment]
@mponton thank you very much for the debugging and tracking down this issue. I'm sure your analysis is correct around the use of those global variables.

@Prativa20 has been working on cleaning up the code and rewriting it on this [branch](https://github.com/vmware/terraform-provider-vra7/tree/no_sling). To your solution #1, please see this [code](https://github.com/vmware/terraform-provider-vra7/blob/no_sling/vrealize/resource.go#L552) which implement it. In an initial test we believe it fixes this issue. Please feel free to try it out and we are looking to merge it to master very soon.

**Comment by mponton** [comment]
@markpeek This is great! I will give it a go later this week, thanks!

**Comment by markpeek** [comment]
FYI...just merged #119 into master. Please verify and report back to this issue when you have time. Thanks!

**Comment by mponton** [comment]
@markpeek Tried to test this tonight but hit another issue (#129) that stopped me (well, I could surely work around it but ran out of time troubleshooting this new one). Will try to check this again tomorrow.

/CC @Prativa20 @jasonlivie

**Comment by mponton** [comment]
OK, so this morning I used `catalog_id` to get around issue #129 and requested 2 VMs in parallel and the fix seems to work. I will ask @jasonlivie to confirm on his side as well. But looking good! Thanks a lot for this @Prativa20 and @markpeek.

**Comment by jasonlivie** [comment]
I confirm that I am now able to perform builds in parallel. This last merge did the trick. Thanks to @Prativa20 @markpeek @mponton for your help.

**Comment by jonnyborbs** [comment]
Awesome news. Thanks for confirming
